Master+optee overlay

bryanodonoghue opened the pull request:
This pull request adds the ability for OPTEE to provide a DTB overlay to a subsequent boot stage.
We can either

1. Append to an existing DTB overlay located at CFG_DT_ADDR or passed in arg2
2. Generate a new DTB overlay at CFG_DT_ADDR

Providing a DT overlay means it is possible for OPTEE to provide internal dynamic data without requiring the DTB to already be in-memory.

Here is an example of merging an OPTEE provided device tree overlay into a main DTB image.

CFG_DT_ADDR = 0x82000000
CFG_DT_OVERLAY = y

u-boot:

=> load mmc 0:1 0x83000000 imx7s-warp.dtb
38564 bytes read in 10 ms (3.7 MiB/s)
=> fdt addr 0x83000000
=> fdt resize 0x1000

=> fdt print /firmware
libfdt fdt_path_offset() returned FDT_ERR_NOTFOUND

=> fdt print /psci
libfdt fdt_path_offset() returned FDT_ERR_NOTFOUND

=> fdt print /reserved-memory
```
reserved-memory {
        #address-cells = <0x00000001>;
        #size-cells = <0x00000001>;
        ranges;
        linux,cma {
                compatible = "shared-dma-pool";
                reusable;
                size = <0x02800000>;
                linux,cma-default;
        };
};

```
Now merge the DTB overlay @ 0x82000000

=> fdt apply 0x82000000
=> fdt print /firmware
```
firmware {
        optee {
                compatible = "linaro,optee-tz";
                method = "smc";
        };
```

=> fdt print /psci
```
psci {
        compatible = "arm,psci-1.0", "arm,psci-0.2", "arm,psci";
        method = "smc";
        cpu_suspend = <0x84000001>;
        cpu_off = <0x84000002>;
        cpu_on = <0x84000003>;
        sys_poweroff = <0x84000008>;
        sys_reset = <0x84000009>;
};
```
=> fdt print /reserved-memory
```
reserved-memory {
        #address-cells = <0x00000002>;
        #size-cells = <0x00000002>;
        ranges;
        optee@0x9fe00000 {
                reg = <0x00000000 0x9fe00000 0x00000000 0x00200000>;
                no-map;
        };
        linux,cma {
                compatible = "shared-dma-pool";
                reusable;
                size = <0x02800000>;
                linux,cma-default;
        };
};

```
The raw in-memory OPTEE DTB overlay looks like this:

=> fdt addr 0x82000000
=> fdt print /
```
/ {
        fragment@2 {
                target-path = "/";
                __overlay__ {
                        reserved-memory {
                                ranges;
                                #size-cells = <0x00000002>;
                                #address-cells = <0x00000002>;
                                optee@0x9fe00000 {
                                        no-map;
                                        reg = <0x00000000 0x9fe00000 0x00000000 0x00200000>;
                                };
                        };
                };
        };
        fragment@1 {
                target-path = "/";
                __overlay__ {
                        psci {
                                sys_reset = <0x84000009>;
                                sys_poweroff = <0x84000008>;
                                cpu_on = <0x84000003>;
                                cpu_off = <0x84000002>;
                                cpu_suspend = <0x84000001>;
                                method = "smc";
                                compatible = "arm,psci-1.0", "arm,psci-0.2", "arm,psci";
                        };
                };
        };
        fragment@0 {
                target-path = "/";
                __overlay__ {
                        firmware {
                                optee {
                                        method = "smc";
                                        compatible = "linaro,optee-tz";
                                };
                        };
                };
        };
};
```
Here is a trivial example of a DTB overlay passed from ATF that has been appended to.

=> fdt addr 0x82000000
=> fdt print /
```
/ {
        fragment@3 {
                target-path = "/";
                __overlay__ {
                        reserved-memory {
                                ranges;
                                #size-cells = <0x00000002>;
                                #address-cells = <0x00000002>;
                                optee@0x9fe00000 {
                                        no-map;
                                        reg = <0x00000000 0x9fe00000 0x00000000 0x00200000>;
                                };
                        };
                };
        };
        fragment@2 {
                target-path = "/";
                __overlay__ {
                        psci {
                                sys_reset = <0x84000009>;
                                sys_poweroff = <0x84000008>;
                                cpu_on = <0x84000003>;
                                cpu_off = <0x84000002>;
                                cpu_suspend = <0x84000001>;
                                method = "smc";
                                compatible = "arm,psci-1.0", "arm,psci-0.2", "arm,psci";
                        };
                };
        };
        fragment@1 {
                target-path = "/";
                __overlay__ {
                        firmware {
                                optee {
                                        method = "smc";
                                        compatible = "linaro,optee-tz";
                                };
                        };
                };
        };
        fragment@0 {
                target = "/";
                __overlay__ {
                        test-int-property = <0x0000002b>;
                };
        };
};

```

jenswi-linaro commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -484,6 +484,56 @@ static void reset_dt_references(void)
 }
 
+#ifdef CFG_DT_OVERLAY
+static int __dt_frag_id;
```
The value of this variable is only valid together with `dt_blob_addr`
To make the connection more clear I think they should be declared together.

Ideally, if it works I'd like to keep them in a struct together. Instead of passing the fdt pointer to `add_optee_dt_node()` and `config_psci()` a pointer to this new struct could be passed.

bryanodonoghue replied:
OK.
I added this change on as a separate patch at the end.

jforissier commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -484,6 +484,56 @@ static void reset_dt_references(void)
 	dt_blob_addr = NULL;
 }
 
+#ifdef CFG_DT_OVERLAY
```
I think this new CFG_ symbol deserves to be documented in `mk/config.mk`. Perhaps add `CFG_DT_OVERLAY ?= n` and some comments after `CFG_DTB_MAX_SIZE`.

jforissier commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -669,6 +725,9 @@ static int add_res_mem_dt_node(void *fdt, const char *name, paddr_t pa,
 		offs = fdt_path_offset(fdt, "/");
+		if (offs < 0)
+			return -1;
+		offs = add_dt_overlay_fragment(fdt, offs);
```
How about adding a helper function for this pattern? (get "/" then add overlay fragment). It is the third time it is used in this file.

etienne-lms commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -484,6 +484,56 @@ static void reset_dt_references(void)
+	int fragment;
+	int ret;
+
+	/* A valid DTB has been passed as a parameter */
```
can remove this comment

etienne-lms commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -506,11 +506,32 @@ static int add_dt_overlay_fragment(void *fdt, int offs)
+	}
+
+	/* No DTB passed - create a new tree */
+	return fdt_create_empty_tree(fdt, fdtsize);
```
An invalid DTB should rather trig an error (`panic()`) than be treated as if early boot did not provide a DTB.

Maybe call `fdt_create_empty_tree(fdt, fdtsize);` only upon `CFG_DT_ADDR` definition as stated in the commit comment.

bryanodonoghue replied:
@etienne-lms what I want to support here is not having to pass an overlay from ATF at all.

bryanodonoghue replied:
edit:
"Maybe call fdt_create_empty_tree(fdt, fdtsize); only upon CFG_DT_ADDR definition as stated in the commit comment."
Yes, OK that works for both objectives, I'll do that.

etienne-lms replied:
I wonder if this is the right way.
Here a DTB with invalid data is silently ignored instead of raising some error.
And if the expect DTB area was not initialized,  `fdt_check_header()` may not successfully fail.

I hardly see how platform can guess if `CFG_DT_ADDR` points to a valid DTB, a existing DTB overlay or an uninitialized memory area.

bryanodonoghue commented:
Changelog
* Rebased ontop of cac9f265
* Added helper functions as suggested by Jérôme
* Added documentation to makefile and documentation/ for overlay as suggested by Jérôme
* Ensure panic() path taken when CFG_DT_OVERLAY is true, bogus pointer passed from ATF and no CFG_DT_ADDR is defined as suggested by Etienne
* Bundled dt blob and frag_id into a single structure as suggested by Jens

etienne-lms commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -485,6 +488,34 @@ static void reset_dt_references(void)
+	int ret;
+
+	snprintf(frag, sizeof(frag), "fragment@%d", __dt_frag_id);
+	offs = fdt_add_subnode(fdt, offs, frag);
```
minor: I think a local variable is preferred than modifying an input argument.

etienne-lms commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -485,6 +488,34 @@ static void reset_dt_references(void)
+		return -1;
+
+	offs = fdt_add_subnode(fdt, offs, "__overlay__");
+	return offs;
```
minor: `return fdt_add_subnode(fdt, offs, "__overlay__");`

etienne-lms commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -485,6 +488,34 @@ static void reset_dt_references(void)
+	offs = fdt_add_subnode(fdt, offs, "__overlay__");
+	return offs;
+}
+
```
minor: remove empty line

etienne-lms commented on mk/config.mk:
```diff
@@ -275,6 +275,16 @@ CFG_DT ?= n
 CFG_DTB_MAX_SIZE ?= 0x10000
 
+# Device Tree Overlay support.
+# This define enables support for an OPTEE provided a DTB overlay.
```
rephrase '... for an OPTEE provided a DTB overlay'. (maybe remove only `a`)

etienne-lms commented on mk/config.mk:
```diff
@@ -275,6 +275,16 @@ CFG_DT ?= n
+#    passed in arg2
+# 2. Generate a new DTB overlay at CFG_DT_ADDR
+# A subsequent boot stage must then merge the generated overlay DTB into a main
+# DTB using the standard fdt_overlay_apply() method.
```
If FDT address boot argument  (`r2` or `CFG_DT_ADDR`) provides a full DTB, not an overlay DTB, should OP-TEE call `fdt_overlay_apply()` for its added fragments to be taken into account?

bryanodonoghue replied:
I don't currently propose supporting that

1. Provide a full DTB and add nodes directly
2. Provide an overlay from OPTEE 
3. Append OPTEE overlay nodes to an existing overlay passed as a parameter

etienne-lms replied:
ok. I wonder if OP-TEE should check whether the provided DTB blob is a DTB or a DTB overlay, or an uninited buffer, prior inserting the OP-TEE fragments. Do you expect `CFG_DT_OVERLAY=y` to implicitly expect DT overlays blobs  only from `r2` boot arg or `CFG_DT_ADDR`?

etienne-lms replied:
Ok, I agree simpler is better. I suggest then to rename the config to something like `CFG_EXTERNAL_DTB_OVERLAY` or `CFG_EXTERNAL_DTB_OVERLAYS`, as opposed to `CFG_EMBEDDED_DT*` stuff.

bryanodonoghue replied:
Fair enough CFG_EXTERNAL_DTB_OVERLAY

etienne-lms commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -523,6 +523,9 @@ static int add_dt_path_subnode(void *fdt, const char *path, const char *subnode)
 		return -1;
+	offs = add_dt_overlay_fragment(fdt, offs);
+		if (offs < 0)
+			return -1;
```
indentation

etienne-lms commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -841,6 +864,22 @@ static void init_fdt(unsigned long phys_fdt)
+	 * The model here is either:
+	 * 1. Pass a DTB overlay from a previous boot stage for appendation
+	 * 2. Pass no DTB overlay. An overlay tree will be created @ CFG_DT_ADDR
+	 * If CFG_DT_OVERLAY is not defined the call to init_dt_overlay is a nop
```
I think you can remove the later comment. Maybe even this full comment.

etienne-lms commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -841,6 +864,22 @@ static void init_fdt(unsigned long phys_fdt)
+	}
+
+	/*
+	 * fdt should point either to a full DTB or a DTB overlay
```
can remove this comment.

bryanodonoghue commented:
Changelog:

CFG_DT_OVERLAY -> CFG_EXTERNAL_DTB_OVERLAY - as suggested-by @etienne-lms

jenswi-linaro commented:
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

etienne-lms commented on documentation/optee_design.md:
```diff
@@ -839,6 +839,28 @@ by early boot and passed to non-secure world are the following:
 
+## Early boot device tree overlay
+
+It is possible to provide a device tree overlay which contains the Early
```
I think those 2 lines should be removed.

bryanodonoghue replied:
I assume you mean
```
It is possible to provide a device tree overlay which contains the Early
boot device tree arguments as a standard device tree overlay structure.
```

etienne-lms commented on documentation/optee_design.md:
```diff
@@ -839,6 +839,28 @@ by early boot and passed to non-secure world are the following:
+
+* Generate a new DTB overlay at CFG_DT_ADDR
+
+A subsequent boot stage must the merge the OPTEE DTB overlay into another
```
suggest to replace `must` with `can` or `should`.

etienne-lms commented on documentation/optee_design.md:
```diff
@@ -839,6 +839,28 @@ by early boot and passed to non-secure world are the following:
+
+There are two possibilities for providing a device tree overlay
+
+* Append OPTEE nodes to an existing DTB overlay located at CFG_DT_ADDR or
```
s/OPTEE/OP-TEE/g  (here and there)

etienne-lms commented on documentation/optee_design.md:
```diff
@@ -839,6 +839,28 @@ by early boot and passed to non-secure world are the following:
+It is possible to provide a device tree overlay which contains the Early
+boot device tree arguments as a standard device tree overlay structure.
+
+There are two possibilities for providing a device tree overlay
```
Proposed rephrasing: 

There are two possibilities for OP-TEE core to provide a device tree overlay to the non-secure world.

etienne-lms commented on documentation/optee_design.md:
```diff
@@ -839,6 +839,28 @@ by early boot and passed to non-secure world are the following:
+ATF -> OPTEE -> u-boot with u-boot merging the OPTEE DTB overlay located at
+CFG_DT_ADDR into a DTB u-boot has loaded from elsewhere.
+
+This functionality is enabled when `CFG_EXTERNAL_DTB_OVERLAY=y`
```
append a terminal dot `.`

etienne-lms commented on core/arch/arm/kernel/generic_boot.c:
```diff
@@ -803,37 +872,51 @@ static void init_fdt(unsigned long phys_fdt)
+	 * 2. Pass no DTB overlay. An overlay tree will be created @ CFG_DT_ADDR
+	 * If CFG_EXTERNAL_DTB_OVERLAY is not defined the call to
+	 * init_dt_overlay is a nop
+	 */
```
this is also  described in the **documentation/optee_design.md** and in **config.mk**.
Should better describe at one place.

bryanodonoghue replied:
I'll zap the comment in the code.

bryanodonoghue commented:
Changelog
- Removed comment from core/arch/arm/kernel/generic_boot.c
  `-       /*
-        * The model here is either:
-        * 1. Pass a DTB overlay from a previous boot stage for appendation
-        * 2. Pass no DTB overlay. An overlay tree will be created @ CFG_DT_ADDR
-        * If CFG_EXTERNAL_DTB_OVERLAY is not defined the call to
-        * init_dt_overlay is a nop
-        */`
- Added Reviewed-by and Acked-by as indicated by @jforissier and @jenswi-linaro  save for the above patch which changed from the last cycle
- Updated remainder of comment observations by @etienne-lms

etienne-lms commented on mk/config.mk:
```diff
@@ -275,6 +275,16 @@ CFG_DT ?= n
+# Device Tree Overlay support.
+# This define enables support for an OPTEE provided DTB overlay.
+# One of two modes is supported in this case:
+# 1. Append OPTEE nodes to an existing DTB overlay located at CFG_DT_ADDR or
```
s/OPTEE/OP-TEE/g   (2 occurrences)

Sorry for such details. config.mk already contain OP-TEE references and your change is better if consistent with the whole file.

jforissier commented:
@bryanodonoghue you may add my
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`
to the third patch ("core: generic_boot: Add init_dt_overlay()").

jforissier commented:
@bryanodonoghue thanks
